File: app/src/main/java/com/app/ralaunch/controls/editor/manager/ControlEditDialogKeymapManager.java
package com.app.ralaunch.controls.editor.manager;

import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.SwitchCompat;

import com.app.ralaunch.R;
import com.app.ralaunch.controls.ControlData;
import com.app.ralaunch.controls.KeyMapper;
import com.app.ralaunch.controls.editor.ControlEditDialogMD;
import com.app.ralaunch.controls.editor.JoystickKeyMappingDialog;
import com.app.ralaunch.controls.editor.KeySelectorDialog;
import com.app.ralaunch.controls.editor.manager.ControlEditDialogVisibilityManager;

/**
 * 键值设置管理器
 * 统一管理普通按钮和摇杆的键值设置逻辑
 */
public class ControlEditDialogKeymapManager {
    
    /**
     * UI元素引用接口
     */
    public interface UIReferences {
        ControlData getCurrentData();
        void notifyUpdate();
    }
    
    /**
     * 绑定键值设置视图
     */
    public static void bindKeymapViews(@NonNull View view, 
                                        @NonNull UIReferences refs,
                                        @NonNull ControlEditDialogMD dialog) {
        View itemKeyMapping = view.findViewById(R.id.item_key_mapping);
        TextView tvKeyName = view.findViewById(R.id.tv_key_name);
        SwitchCompat switchToggleMode = view.findViewById(R.id.switch_toggle_mode);
        
        // 普通按钮的键值映射
        if (itemKeyMapping != null) {
            itemKeyMapping.setOnClickListener(v -> showKeySelectDialog(dialog, refs, tvKeyName));
        }
        
        // 摇杆的键值映射
        View itemJoystickKeyMapping = view.findViewById(R.id.item_joystick_key_mapping);
        TextView tvJoystickKeyMappingValue = view.findViewById(R.id.tv_joystick_key_mapping_value);
        if (itemJoystickKeyMapping != null) {
            itemJoystickKeyMapping.setOnClickListener(v -> showJoystickKeyMappingDialog(dialog, refs, tvJoystickKeyMappingValue));
        }
        
        if (switchToggleMode != null) {
            switchToggleMode.setOnCheckedChangeListener((buttonView, isChecked) -> {
                if (refs.getCurrentData() != null) {
                    refs.getCurrentData().isToggle = isChecked;
                    refs.notifyUpdate();
                }
            });
        }
    }
    
    /**
     * 显示摇杆键值映射对话框
     */
    private static void showJoystickKeyMappingDialog(@NonNull ControlEditDialogMD dialog,
                                                     @NonNull UIReferences refs,
                                                     TextView tvJoystickKeyMappingValue) {
        ControlData data = refs.getCurrentData();
        if (data == null || data.type != ControlData.TYPE_JOYSTICK) {
            return;
        }
        
        // 确保 joystickKeys 存在
        if (data.joystickKeys == null || data.joystickKeys.length < 4) {
            data.joystickKeys = new int[]{
                ControlData.SDL_SCANCODE_W,  // up
                ControlData.SDL_SCANCODE_D,  // right
                ControlData.SDL_SCANCODE_S,  // down
                ControlData.SDL_SCANCODE_A   // left
            };
        }
        
        JoystickKeyMappingDialog keyMappingDialog = new JoystickKeyMappingDialog(
            dialog.getContext(), data, (updatedData) -> {
                // 更新数据
                data.joystickKeys = updatedData.joystickKeys;
                
                // 更新显示
                updateJoystickKeyMappingDisplay(tvJoystickKeyMappingValue, data);
                
                refs.notifyUpdate();
            });
        
        keyMappingDialog.show();
    }
    
    /**
     * 更新摇杆键值映射显示
     */
    private static void updateJoystickKeyMappingDisplay(TextView tv, ControlData data) {
        if (tv == null || data == null || data.joystickKeys == null || data.joystickKeys.length < 4) {
            return;
        }
        
        String up = KeyMapper.getKeyName(data.joystickKeys[0]);
        String right = KeyMapper.getKeyName(data.joystickKeys[1]);
        String down = KeyMapper.getKeyName(data.joystickKeys[2]);
        String left = KeyMapper.getKeyName(data.joystickKeys[3]);
        
        String displayText = tv.getContext().getString(R.string.editor_joystick_key_mapping_value, up, right, down, left);
        tv.setText(displayText);
    }
    
    /**
     * 显示按键选择对话框
     */
    private static void showKeySelectDialog(@NonNull ControlEditDialogMD dialog,
                                            @NonNull UIReferences refs,
                                            TextView tvKeyName) {
        ControlData data = refs.getCurrentData();
        if (data == null) return;
        
        boolean isGamepadMode = (data.buttonMode == ControlData.BUTTON_MODE_GAMEPAD);
        KeySelectorDialog keyDialog = new KeySelectorDialog(dialog.getContext(), isGamepadMode);
        
        keyDialog.setOnKeySelectedListener((keycode, keyName) -> {
            data.keycode = keycode;
            // 使用 KeyMapper 获取完整的按键名称，确保显示正确
            String fullKeyName = KeyMapper.getKeyName(keycode);
            if (tvKeyName != null) {
                tvKeyName.setText(fullKeyName);
            }
            refs.notifyUpdate();
        });
        
        keyDialog.show();
    }
    
    /**
     * 更新键值设置视图的可见性
     */
    public static void updateKeymapVisibility(@NonNull View keymapView, @NonNull ControlData data) {
        View itemKeyMapping = keymapView.findViewById(R.id.item_key_mapping);
        View itemJoystickKeyMapping = keymapView.findViewById(R.id.item_joystick_key_mapping);
        View itemToggleMode = keymapView.findViewById(R.id.item_toggle_mode);
        TextView tvJoystickKeyMappingValue = keymapView.findViewById(R.id.tv_joystick_key_mapping_value);
        
        // 普通按钮：显示单个按键映射（文本控件和摇杆不支持单个按键映射）
        boolean isButton = (data.type == ControlData.TYPE_BUTTON);
        boolean isText = (data.type == ControlData.TYPE_TEXT);
        boolean isJoystick = (data.type == ControlData.TYPE_JOYSTICK);
        
        // 普通按钮的键值映射：只有按钮类型（非文本、非摇杆）才显示
        boolean shouldShowButtonKeymap = isButton && !isText && !isJoystick;
        
        // 摇杆的键值映射：只有摇杆类型且为键盘模式才显示
        boolean shouldShowJoystickKeymap = isJoystick && 
            data.joystickMode == ControlData.JOYSTICK_MODE_KEYBOARD;
        
        if (itemKeyMapping != null) {
            itemKeyMapping.setVisibility(shouldShowButtonKeymap ? View.VISIBLE : View.GONE);
        }
        if (itemJoystickKeyMapping != null) {
            itemJoystickKeyMapping.setVisibility(shouldShowJoystickKeymap ? View.VISIBLE : View.GONE);
            // 更新显示值
            if (shouldShowJoystickKeymap && tvJoystickKeyMappingValue != null) {
                updateJoystickKeyMappingDisplay(tvJoystickKeyMappingValue, data);
            }
        }
        if (itemToggleMode != null) {
            itemToggleMode.setVisibility(shouldShowButtonKeymap ? View.VISIBLE : View.GONE);
        }
    }
}

File: app/src/main/java/com/app/ralaunch/controls/editors/managers/ControlEditDialogKeymapManager.kt
package com.app.ralaunch.controls.editors.managers

import android.view.View
import android.widget.TextView
import androidx.appcompat.widget.SwitchCompat
import com.app.ralaunch.R
import com.app.ralaunch.controls.KeyMapper
import com.app.ralaunch.controls.configs.ControlData
import com.app.ralaunch.controls.configs.ControlData.Joystick
import com.app.ralaunch.controls.configs.ControlData.KeyCode
import com.app.ralaunch.controls.editors.ControlEditDialogMD
import com.app.ralaunch.controls.editors.JoystickKeyMappingDialog
import com.app.ralaunch.controls.editors.KeySelectorDialog

/**
 * 键值设置管理器
 * 统一管理普通按钮和摇杆的键值设置逻辑
 */
object ControlEditDialogKeymapManager {
    /**
     * 绑定键值设置视图
     */
    fun bindKeymapViews(
        view: View,
        refs: UIReferences,
        dialog: ControlEditDialogMD
    ) {
        val itemKeyMapping = view.findViewById<View?>(R.id.item_key_mapping)
        val tvKeyName = view.findViewById<TextView?>(R.id.tv_key_name)
        val switchToggleMode = view.findViewById<SwitchCompat?>(R.id.switch_toggle_mode)

        // 普通按钮的键值映射
        if (itemKeyMapping != null) {
            itemKeyMapping.setOnClickListener {
                showKeySelectDialog(
                    dialog,
                    refs,
                    tvKeyName
                )
            }
        }

        // 摇杆的键值映射
        val itemJoystickKeyMapping = view.findViewById<View?>(R.id.item_joystick_key_mapping)
        val tvJoystickKeyMappingValue =
            view.findViewById<TextView?>(R.id.tv_joystick_key_mapping_value)
        if (itemJoystickKeyMapping != null) {
            itemJoystickKeyMapping.setOnClickListener {
                showJoystickKeyMappingDialog(
                    dialog,
                    refs,
                    tvJoystickKeyMappingValue
                )
            }
        }

        if (switchToggleMode != null) {
            switchToggleMode.setOnCheckedChangeListener { _, isChecked ->
                val data = refs.currentData
                if (data is ControlData.Button) {
                    data.isToggle = isChecked
                    refs.notifyUpdate()
                }
            }
        }
    }

    /**
     * 显示摇杆键值映射对话框
     */
    private fun showJoystickKeyMappingDialog(
        dialog: ControlEditDialogMD,
        refs: UIReferences,
        tvJoystickKeyMappingValue: TextView?
    ) {
        val data = refs.currentData

        if (data !is Joystick) {
            return
        }

        val joystickData = data

        // 确保摇杆键值数组已初始化
        if (joystickData.joystickKeys.isEmpty()) {
            joystickData.joystickKeys = arrayOf(
                KeyCode.KEYBOARD_W,  // up
                KeyCode.KEYBOARD_D,  // right
                KeyCode.KEYBOARD_S,  // down
                KeyCode.KEYBOARD_A // left
            )
        }

        val keyMappingDialog = JoystickKeyMappingDialog(
            dialog.requireContext(), joystickData,
            object : JoystickKeyMappingDialog.OnSaveListener {
                override fun onSave(data: ControlData?) {
                    // 更新键值数据
                    if (data is Joystick) {
                        joystickData.joystickKeys = data.joystickKeys

                        // 更新显示
                        updateJoystickKeyMappingDisplay(tvJoystickKeyMappingValue, joystickData)
                        refs.notifyUpdate()
                    }
                }
            })
        keyMappingDialog.show()
    }

    /**
     * 更新摇杆键值映射显示
     */
    private fun updateJoystickKeyMappingDisplay(tv: TextView?, data: ControlData?) {
        if (tv == null || data !is Joystick) {
            return
        }

        val joystickData = data
        val keys: Array<KeyCode> = joystickData.joystickKeys

        if (keys.size < 4) {
            return
        }

        val keyMapper = KeyMapper
        val up = keyMapper.getKeyName(keys[0])
        val right = keyMapper.getKeyName(keys[1])
        val down = keyMapper.getKeyName(keys[2])
        val left = keyMapper.getKeyName(keys[3])
        tv.text = "↑$up ↓$down ←$left →$right"
    }

    /**
     * 显示按键选择对话框
     */
    private fun showKeySelectDialog(
        dialog: ControlEditDialogMD,
        refs: UIReferences,
        tvKeyName: TextView?
    ) {
        val data = refs.currentData
        if (data == null) {
            return
        }

        var isGamepadMode = false
        if (data is ControlData.Button) {
            isGamepadMode = data.mode == ControlData.Button.Mode.GAMEPAD
        }

        val keyDialog = KeySelectorDialog(dialog.requireContext(), isGamepadMode)

        keyDialog.setOnKeySelectedListener(object : KeySelectorDialog.OnKeySelectedListener {
            override fun onKeySelected(keycode: Int, keyName: String?) {
                if (data is ControlData.Button) {
                    val buttonData = data

                    // Convert int keycode to KeyCode enum
                    val keycodeEnum = findKeyCodeByValue(keycode)
                    buttonData.keycode = keycodeEnum

                    val keyMapper = KeyMapper
                    val fullKeyName = keyMapper.getKeyName(keycodeEnum)
                    if (tvKeyName != null) {
                        tvKeyName.text = fullKeyName
                    }
                    refs.notifyUpdate()
                }
            }
        })
        keyDialog.show()
    }

    /**
     * 根据int值查找KeyCode枚举
     */
    private fun findKeyCodeByValue(value: Int): KeyCode {
        for (keyCode in KeyCode.entries) {
            if (keyCode.code == value) {
                return keyCode
            }
        }
        return KeyCode.UNKNOWN
    }

    /**
     * 更新键值设置选项的可见性（根据控件类型）
     */
    fun updateKeymapVisibility(keymapView: View, data: ControlData) {
        val itemKeyMapping = keymapView.findViewById<View?>(R.id.item_key_mapping)
        val itemJoystickKeyMapping = keymapView.findViewById<View?>(R.id.item_joystick_key_mapping)
        val itemToggleMode = keymapView.findViewById<View?>(R.id.item_toggle_mode)

        // 按钮和文本控件显示普通键值设置
        val isButton = data is ControlData.Button
        val isText = data is ControlData.Text
        val isJoystick = data is Joystick

        // 普通键值设置（仅按钮和文本控件显示）
        if (itemKeyMapping != null) {
            itemKeyMapping.setVisibility(if (isButton || isText) View.VISIBLE else View.GONE)
        }

        // 摇杆键值设置（仅摇杆显示，且为键盘模式）
        var showJoystickKeyMapping = false
        if (isJoystick) {
            val joystickData = data
            showJoystickKeyMapping = joystickData.mode == Joystick.Mode.KEYBOARD
        }

        if (itemJoystickKeyMapping != null) {
            itemJoystickKeyMapping.setVisibility(if (showJoystickKeyMapping) View.VISIBLE else View.GONE)
        }

        // 切换模式（仅按钮显示）
        if (itemToggleMode != null) {
            itemToggleMode.setVisibility(if (isButton) View.VISIBLE else View.GONE)
        }
    }

    /**
     * UI元素引用接口
     */
    interface UIReferences {
        val currentData: ControlData?
        fun notifyUpdate()
    }
}

